File: mcp-code-intelligence-python/src/mcp_code_intel/memory/decision.py
"""DecisionMemory — tracks architectural decisions and their rationale."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass
class Decision:
    """A structured decision record (ADR)."""

    title: str
    context: str
    decision: str
    rationale: str
    alternatives: list[str] = field(default_factory=list)
    consequences: str = ""
    source: str | None = None
    tags: str = ""


class DecisionMemory:
    """Records and retrieves architectural decisions."""

    def __init__(self, repo: Any, graph_repo: Any) -> None:
        self._repo = repo
        self._graph = graph_repo

    def record_decision(self, decision: Decision) -> int:
        """Record a new decision."""
        content = self._format_content(decision)
        return self._repo.insert_entry(
            content=content,
            summary=f"Decision: {decision.title}",
            entry_type="DECISION",
            tier="EPISODIC",
            source=decision.source,
            tags=decision.tags,
            confidence=0.9,
        )

    def link_decision(self, decision_id: int, related_id: int, relation: str = "RELATES_TO") -> None:
        """Link a decision to related entries."""
        self._graph.add_edge(decision_id, related_id, relation)

    def find_decisions(self, limit: int = 20) -> list[dict]:
        """Find decisions."""
        return self._repo.find_by_type("DECISION", limit)

    def _format_content(self, d: Decision) -> str:
        lines: list[str] = []
        lines.append(f"## Context\n{d.context}")
        lines.append(f"\n## Decision\n{d.decision}")
        lines.append(f"\n## Rationale\n{d.rationale}")
        if d.alternatives:
            lines.append("\n## Alternatives Considered")
            lines.extend(f"- {a}" for a in d.alternatives)
        if d.consequences:
            lines.append(f"\n## Consequences\n{d.consequences}")
        return "\n".join(lines)

File: backend/mcp-code-intelligence-nodejs/src/memory/decision.ts
/**
 * DecisionMemory — tracks architectural decisions and their rationale.
 */

export interface Decision {
  title: string;
  context: string;
  decision: string;
  rationale: string;
  alternatives?: string[];
  consequences?: string;
  source?: string;
  tags?: string;
}

export class DecisionMemory {
  private repo: any;
  private graphRepo: any;

  constructor(repo: any, graphRepo: any) {
    this.repo = repo;
    this.graphRepo = graphRepo;
  }

  /** Record a new decision. */
  recordDecision(decision: Decision): number {
    const content = this.formatContent(decision);
    return this.repo.insert({
      content,
      summary: `Decision: ${decision.title}`,
      type: 'DECISION',
      tier: 'EPISODIC',
      source: decision.source ?? null,
      tags: decision.tags ?? '',
      confidence: 0.9,
    });
  }

  /** Link a decision to related entries. */
  linkDecision(decisionId: number, relatedId: number, relation = 'RELATES_TO'): void {
    this.graphRepo.addEdge({ sourceId: decisionId, targetId: relatedId, relation });
  }

  /** Find decisions. */
  findDecisions(limit = 20): any[] {
    return this.repo.findByType('DECISION', limit);
  }

  private formatContent(d: Decision): string {
    const lines: string[] = [];
    lines.push(`## Context\n${d.context}`);
    lines.push(`\n## Decision\n${d.decision}`);
    lines.push(`\n## Rationale\n${d.rationale}`);
    if (d.alternatives && d.alternatives.length > 0) {
      lines.push('\n## Alternatives Considered');
      d.alternatives.forEach(a => lines.push(`- ${a}`));
    }
    if (d.consequences) {
      lines.push(`\n## Consequences\n${d.consequences}`);
    }
    return lines.join('\n');
  }
}
